### PR TITLE
Support Iceberg default warehouse location config

### PR DESCRIPTION
**Pull request by jackye1995**
Support using a configurable warehouse location to determine the default table location.

This is to match the Iceberg side catalog behavior: https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L452-L459.

@losipiuk @findepi

### CODE REVIEW

**Comment by findepi** [comment]
In Iceberg, I can define `location` property on the schema.
@jackye1995  Why do we need this?

cc @losipiuk @phd3 @alexjo2144

**Comment by jackye1995** [comment]
> In Iceberg, I can define location property on the schema. Why do we need this?

There are 2 reasons for this config:

1. Iceberg has the concept of a default warehouse location for catalog, so that users do not have to specify table location. This is similar to Hive's `hive.metastore.warehouse.dir`. But as you know Iceberg does not prefer using Hadoop config, so a warehouse location config is used. Adding this config in Trino brings Iceberg and Trino catalog concept closer.
2. Hadoop catalog actually requires such a config. If user supplies an alternative table location, it should throw an exception. This is because Hadoop catalog relies on the notion of `warehouse/db/table` to store catalog architecture. It does file listing to find all the databases and tables within a database. So this config is necessary for Hadoop catalog implementation.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 1]
```diff
@@ -51,6 +53,19 @@ public IcebergConfig setCatalogType(CatalogType catalogType)
...
 
+    @Config("iceberg.catalog.warehouse")
+    @Description("Iceberg default warehouse location, used to generate default table location")
+    public IcebergConfig setCatalogWarehouse(String warehouse)
+    {
+        this.catalogWarehouse = warehouse;
+        return this;
+    }
+
+    public String getCatalogWarehouse()
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:64]: Make the property `Optional<String>`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java [hunk 2]
```diff
@@ -51,6 +53,19 @@ public IcebergConfig setCatalogType(CatalogType catalogType)
         return this;
     }
 
+    @Config("iceberg.catalog.warehouse")
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:56]: not sure about the name. something like "default storage location"?


cc @losipiuk @alexjo2144 @phd3

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoCatalogFactory.java [hunk 3]
```diff
@@ -39,6 +39,7 @@
     private final CatalogType catalogType;
     private final boolean isUniqueTableLocation;
     private final boolean isUsingSystemSecurity;
+    private final String warehouse;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoCatalogFactory.java:42]: This should be named more meaningfully.  Let's figure config property name first.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 4]
```diff
@@ -319,6 +324,16 @@ public String defaultTableLocation(ConnectorSession session, SchemaTableName sch
         if (useUniqueTableLocation) {
             tableNameForLocation += "-" + randomUUID().toString().replace("-", "");
         }
+        if (database.getLocation().isEmpty() || database.getLocation().get().isEmpty()) {
+            if (warehouse == null) {
+                throw new TrinoException(HIVE_DATABASE_LOCATION_ERROR, format("Database '%s' location cannot be determined, " +
+                        "please either set 'location' when creating the database, or set 'iceberg.catalog.warehouse' " +
+                        "to allow a default location at 'warehousePath/databaseName.db'", database.getDatabaseName()));
+            }
+            database = Database.builder(database)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:333]: Let's not build a fake Database object. 
Instead, let's modify `getTableDefaultLocation` (which already has related logic).

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java [hunk 5]
```diff
@@ -319,6 +324,16 @@ public String defaultTableLocation(ConnectorSession session, SchemaTableName sch
...
+                    .setLocation(Optional.of(format("%s/%s.db", warehouse, schemaTableName.getSchemaName())))
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:334]: why `.db`?

**Review by findepi — COMMENTED** [review]
Add a test covering creation of tables without location, within schema without location set.
This is what becomes possible with this change.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java (hunk 2 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:56]: This is to match https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/CatalogProperties.java#L31. Please let me know if other name is preferred, I can update to that.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoCatalogFactory.java (hunk 3 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoCatalogFactory.java:42]: sure, I updated to `catalogWarehouse` for now

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java (hunk 5 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TrinoHiveCatalog.java:334]: This is to match https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L433-L440, not sure if Trino has any preference for this

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java (hunk 1 above)
Comment (jackye1995) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:64]: updated

**Comment by jackye1995** [comment]
@findepi test is actually quite difficult to add, the file metastore has a hard-coded database location, so that code path would never be exercised. Not sure if there is any other good way to trigger that.

**Comment by jackye1995** [comment]
We can also choose to not have this feature for Hive catalog and only support it with Hadoop and Glue catalog, because based on the current logic of `HiveWriteUtils.getTableDefaultLocation`, table name creation should fail if the provided path does not exist in non-S3 HDFS environment. So it seems like Trino assumes the existence of a database location for this to work for Hive. Any thoughts?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java (hunk 2 above)
Comment (losipiuk) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergConfig.java:64]: For schemas and tables we use "location". So probably we should use it also here. `iceberg.default-schema-location`?

**Comment by jackye1995** [comment]
close in favor of #10151

**Comment by findepi** [comment]
> This is to match the Iceberg side catalog behavior: https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java#L452-L459.

The link no longer seems to be adequate. @jackye1995 can you link to a file in a commit, instead of a branch?

**Pull request closed without merging**